Topics/Development_Process/Build_Requirements/CMakeLists_txt.md

```diff
@@ -0,0 +1,199 @@
+
+The key feature of CMake is that it supports the cross-platform discovery of system libraries, and automatic discovery and configuration of the toolchain, which is much easier to use than Make.
+
+To install CMake, check their [document](https://cmake.org/download/) on how to download and install on Unix/Linux or Windows platform.
+
+
+## What is CMakeLists.txt?
+
+When CMake processes a project, the entry point is a source file called `CMakeLists.txt` in the top level of the source directory.
+
+`CMakeLists.txt` file is the CMake configuration file, which contains a set of directives and instructions describing the project's source files and taargets (executable, library, or both) that will be used to build the program. It determines everything for the building process, from which source files to compile, to which options to choose for the libraries and to present to the users.
+
+## Creating CMakeLists.txt
+
+### CMake Language
+`CMakeLists.txt` is written in [CMake Language](https://cmake.org/cmake/help/latest/manual/cmake-language.7.html#manual:cmake-language(7)), which consists of comments, [commands](https://cmake.org/cmake/help/latest/manual/cmake-commands.7.html#manual:cmake-commands(7)), and [variables](https://cmake.org/cmake/help/latest/manual/cmake-variables.7.html#manual:cmake-variables(7)). 
+
+* **Comments** start with `#`
+  ```
+  # This is a comment
+  ```
+* **Commands** start with command names and a list of whitespace-separated arguments (case-sensitive).
+  ```
+  find_package(LIBIGL REQUIRED QUIET)
+  ```
+* **Variables** are case-sensitive with only alphanumeric characters and underscores. 
+
+  A number of useful variables are automatically defined by CMake (e.g `CMAKE_CURRENT_SOURCE_DIR`, `PROJECT_NAME`). 
+
+  Use *set* command to set variable names, and *${variable_name}* to reference a varible in command arguments. 
+  ```
+  set(Foo a b c)    # Value of Foo is "a;b;c"
+
+  command(${Foo})   # Foo is replaced by a;b;c 
+                    # and expands to 3 arguments
+  ```
+
+It also provides the basic **Flow Control** - Conditional statements, Looping constructs, and Procedure definitions. Here are the examples:
+
+* **Conditional statements (e.g [if](https://cmake.org/cmake/help/latest/command/if.html#command:if))**
+    ```
+    if(<condition>)
+    <commands>
+    elseif(<condition>)     # optional block, can be repeated
+    <commands>
+    else()                  # optional block
+    <commands>
+    endif()
+    ```
+* **Looping Constructs (e.g [foreach](https://cmake.org/cmake/help/latest/command/foreach.html#command:foreach), [while](https://cmake.org/cmake/help/latest/command/while.html#command:while))**
+    ```
+    foreach(<variable>                  # <variable> to be referenced
+            <list of variables>)        # <list of variables> to replace ${variable}
```

Comment:
Took a while for me to realize that this is equivalent to `for <variable> in <list of variables>` in Python -- I think `<list of variables> to replace ${variable}` is a slightly confusing way to word things. Thought it meant "Alternatively, instead of providing one variable, you can provide a list of variables" at first.

Reply:
Thanks for finding this. `for <variable> in <list of variables>` is way much clearer.